pkg/core/index.go

```diff
@@ -0,0 +1,249 @@
+	l              *zap.Logger
+	_              struct{}
+}
+
+func defaultFileIndex(stores context2.Stores) *fileIndex {
+	return &fileIndex{
+		metaObject:     defaultMetaObject(stores.VMetadata()), // TODO(fred): make this more generic for reuse, e.g. with fileIndexOptions
+		output:         make(chan bundleEntriesRes, 2*defaultBundleEntriesPerFile),
+		concurrency:    100,
+		l:              dlogger.MustGetLogger("info"),
+		entriesPerFile: defaultBundleEntriesPerFile,
+	}
+}
+
+func newFileIndex(stores context2.Stores, opts ...fileIndexOption) *fileIndex {
+	f := defaultFileIndex(stores)
+	for _, apply := range opts {
+		apply(f)
+	}
+	return f
+}
+
+// Iterator returns the current patherIterator for this infex
+func (f *fileIndex) Iterator() patherIterator {
+	return f.indexPather
+}
+
+// Download bundle or splits file entries into an output channel.
+//
+// Download may iterate over several bundles or splits.
+func (f *fileIndex) Download() error {
+	return f.unpack()
+}
+
+// OutputChan provides the output channel for Download.
+//
+// Download will close the channel when done.
+func (f *fileIndex) OutputChan() chan bundleEntriesRes {
+	return f.output
+}
+
+// Upload file entries for a single bundle or split from input channel to index files.
+//
+// It returns the number of index files uploaded.
+func (f *fileIndex) Upload(filePackedC chan filePacked, errorC chan errorHit, doneOkC chan struct{}) (uint64, error) {
+	return f.pack(filePackedC, errorC, doneOkC)
+}
+
+// uploadIndex uploads a single file list index file
+func (f *fileIndex) uploadIndex(fileList []model.BundleEntry, pather func(uint64) string, index uint64) error {
+	buffer, err := yaml.Marshal(model.BundleEntries{
+		BundleEntries: fileList,
+	})
+	if err != nil {
+		return err
+	}
+
+	err = f.writeMetadata(pather(index), buffer)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *fileIndex) getFileIndex(pth string) (model.BundleEntries, error) {
+	buffer, err := f.readMetadata(pth)
+	if err != nil {
+		return model.BundleEntries{}, err
+	}
+
+	var entries model.BundleEntries
+	if err := yaml.Unmarshal(buffer, &entries); err != nil {
+		return model.BundleEntries{}, err
+	}
+	return entries, nil
+}
+
+// downloadIndex downloads the i-th file index located by pather
+func (f *fileIndex) downloadIndex(id string, pather func(uint64) string, index uint64, chans downloadBundleFileListChans) {
+	defer func() {
+		<-chans.concurrencyControl
+	}()
+
+	f.l.Info("downloading index file", zap.Uint64("current entry", index))
+	entries, err := f.getFileIndex(pather(index))
+	if err != nil {
+		chans.error <- err
+		return
+	}
+
+	chans.bundleEntries <- bundleEntriesRes{bundleEntries: entries, idx: index, id: id}
+}
+
+// downloadAll downloads all index files in parallel, given an iterator to point to the file location
+func (f *fileIndex) downloadAll(chans downloadBundleFileListChans) {
+	f.indexPather.Reset()
+
+	throttle := make(chan struct{}, f.concurrency)
```

Comment:
i think this is a semaphore?

i'm not clear what "throttle" means as a noun in terms of programming.  i usually think throttle is a verb?

not s'th to get too delve into too much either way.

Reply:
> A throttle is the mechanism by which fluid flow is managed by constriction or obstruction. An engine's power can be increased or decreased by the restriction of inlet gases, but usually decreased.